File: helpers/rust/example/contract.rs
use serde::Deserialize;
use serde::Serialize;
use three_em::handler;

#[derive(Serialize, Deserialize)]
pub struct State {
  counter: i32,
}

#[derive(Deserialize)]
pub struct Action {}

#[handler]
pub fn handle(state: State, _action: Action) -> State {
  State {
    counter: state.counter + 1,
  }
}

File: helpers/cpp/contract.c
#include "three_em.h"

cJSON *
handler (cJSON * state, cJSON * action)
{
  cJSON *result = NULL;
  cJSON *counter = NULL;

  counter = cJSON_GetObjectItemCaseSensitive (state, "counter");
  result = cJSON_CreateObject ();
  cJSON_AddNumberToObject (result, "counter", counter->valueint + 1);
  return result;
}

MAKE_CONTRACT (handler)
